gadget/mountedfilesystem.go

```diff
@@ -179,10 +198,578 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+// and target specification.
+//  source: /foo/bar/ target: /baz => /bar/    (contents of /foo/bar/ under /baz)
+//  source: /foo/bar  target: /baz => /bar/bar (directory /foo/bar under /baz, contents under /baz/bar)
+func targetForSourceDir(source, target string) string {
+	if strings.HasSuffix(source, "/") {
+		// contents of source directory land under target
+		return target
+	}
+	// source directory lands under target
+	return filepath.Join(target, filepath.Base(source))
+}
+
+func (f *MountedFilesystemUpdater) makePrefix(dstRoot, target string) error {
+	dstPath, _ := f.entryDestPaths(dstRoot, "", target, "")
+	if err := os.MkdirAll(filepath.Dir(dstPath), 0755); err != nil {
+		return fmt.Errorf("cannot write prefix directory: %v", err)
+	}
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) updateDirectory(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	fis, err := f.sourceDirectoryEntries(source)
+	if err != nil {
+		return fmt.Errorf("cannot list source directory %q: %v", source, err)
+	}
+
+	target = targetForSourceDir(source, target)
+
+	// create current target directory if needed
+	if err := os.MkdirAll(filepath.Join(dstRoot, target), 0755); err != nil {
+		return fmt.Errorf("cannot write directory: %v", err)
+	}
+	// and write the content of source to target
+	for _, fi := range fis {
+		pSrc := filepath.Join(source, fi.Name())
+		pDst := filepath.Join(target, fi.Name())
+
+		update := f.updateOrSkipFile
+		if fi.IsDir() {
+			// continue updating contents of the directory rather
+			// than the directory itself
+			pSrc += "/"
+			pDst += "/"
+			update = f.updateDirectory
+		}
+		if err := f.makePrefix(dstRoot, target); err != nil {
+			return err
+		}
+		if err := update(dstRoot, pSrc, pDst, preserveInDst, backupDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) updateOrSkipFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcPath, dstPath, backupPath := f.entryPaths(dstRoot, source, target, backupDir)
+
+	if osutil.FileExists(dstPath) {
+		if strutil.SortedListContains(preserveInDst, dstPath) {
+			// file is to be preserved
+			return nil
+		}
+		if osutil.FileExists(backupPath + ".same") {
+			// file is the same as current copy
+			return nil
+		}
+		if !osutil.FileExists(backupPath + ".backup") {
+			// not preserved & different than the update, error out
+			// as there is no backup
+			return fmt.Errorf("missing backup file %v for %v", backupPath+".backup", target)
+		}
+	}
+
+	return writeFile(srcPath, dstPath, preserveInDst)
+}
+
+func (f *MountedFilesystemUpdater) updateVolumeContent(volumeRoot string, content *VolumeContent, preserveInDst []string, backupDir string) error {
+	if err := checkContent(content); err != nil {
+		return err
+	}
+
+	srcPath := f.entrySourcePath(content.Source)
+
+	if err := f.makePrefix(volumeRoot, content.Target); err != nil {
```

Comment:
Is this useful? both writeFile and updateDirectory have MkdirAll. If I comment out this no tests fail.

Reply:
Yeah, after all the changes this became irrelevant. Dropped now.